### PR TITLE
Remove non-pinned packages more thoroughly

### PR DESCRIPTION
**Pull request by ericphanson**
This is easier explained with an example. Let's say I have AWSS3.jl and AWS.jl in my Project.toml, and I was working on AWS.jl so I un-pinned it so that I don't build it into my sysimage, because I want to be able to change versions (e.g. switch to my branch) without rebuilding it. Currently, we would `rm` AWS.jl from the Project before building the sysimage. But we would *not* rm AWSS3, which means that AWSS3 and *all of its dependencies* including AWS.jl would get baked into the sysimage. That means I would not be able to change versions of AWS.jl despite unpinning it.

Here, we use a Pkg command to remove the packages everywhere in the Manifest. That means if any package depends on a removed package, it is removed too. So in the example, we would not build either AWSS3 or AWS into the sysimage; they would both be removed since AWS.jl is not pinned.

That means if you do not pin something very low in the stack, you might end up removing most things from your sysimage. But that's basically the cost of editing something low in the stack I think; it seems unavoidable.

### CODE REVIEW

**Comment by ericphanson** [comment]
Ah, turns out this removes non-pinned stdlibs too, and anything depending on them. Hm..

**Review by kolia — APPROVED** [review]
Nice!

TIL about `Pkg.PKGMODE_MANIFEST`. The `Pkg.rm` docs say

```If the mode of pkg is PKGMODE_MANIFEST also remove it from the manifest including all recursive dependencies of pkg.```

which is confusing to me, it's actually removing packages that depend on `pkg`, I read `dependencies of pkg` to mean going up the dependency DAG in the other direction.

**Comment by kolia** [comment]
Yeah, this does assume that users typically pin a maximal set of dependencies. If you forget some, it may `rm` a whole bunch of others. But I guess that's ok.

**Comment by ericphanson** [comment]
Yeah, I think that's a bit unfortunate. But I think maybe it's worse if there's confusing bugs over not realizing that a package got baked into the sysimage and you don't get why it's not seeing the new version.

**Comment by ericphanson** [comment]
I updated the logic a bit: now a package is set to be removed (recursively as discussed) if:
* it is a direct dependency (unpinned manifest deps are allowed to stay)
* it is not a stdlib (stdlibs are presumed pinned)
* it is not pinned

and to allow the "everything is pinned" scenario. (`Pkg.rm` errors if you give it an empty list).

**Comment by kolia** [comment]
Conversely, I've been wondering whether we should pin Manifest deps that are transitive dependencies of a pinned package. Right now without that, your sysimage has these transitive dependencies baked in, but Pkg doesn't know that and will happily think it's for example upgrading one of these when in fact it is fixed in sysimage stone.

The reason I haven't done that is that it doesn't seem too polite to modify the user's set of pinned packages for them. Could throw an error and provide some convenience function for pinning transitive deps, but that's not great either.

Also not sure how much trouble not doing will cause in practice, I suspect not much.

**Comment by ericphanson** [comment]
Yeah... I don't know what we should do here with respect to that.

@KristofferC has suggested on Slack that an upstream fix would be for PackageCompiler to record the versions of all the packages baked into the sysimage, and to teach Pkg.jl to read that and treat those packages as pinned.

That way Pkg wouldn't lie about upgrades and such. That would be pretty nice!

File: add_me_to_your_PATH/Dockerfile.template [hunk 1]
```diff
@@ -7,10 +7,14 @@ FROM julia:${JULIA_VERSION}-buster as pinned-extractor
...
+# Remove all packages that are not pinned recursively from the manifest
+# (stdlibs count as pinned)
 RUN julia --project=. -e 'using Pkg;\
-                          pinned = [p.name for p in values(Pkg.dependencies()) if p.is_pinned];\
-                          Pkg.rm([p for p in keys(Pkg.project().dependencies) if p ∉ pinned])'
+                          not_pinned = [p.name for p in values(Pkg.dependencies()) if\
+                                        p.is_direct_dep &&\
+                                        !p.is_pinned &&\
+                                        p.name ∉ values(Pkg.Types.stdlibs())];\
+                          isempty(not_pinned) || Pkg.rm(not_pinned; mode=Pkg.PKGMODE_MANIFEST)'
```
Comment (omus) [inline add_me_to_your_PATH/Dockerfile.template:17]: Surface level changes:
```suggestion
                          unpinned = [p.name for p in values(Pkg.dependencies()) \
                                      if p.is_direct_dep && !p.is_pinned && \
                                      p.name ∉ values(Pkg.Types.stdlibs())]; \
                          isempty(unpinned) || Pkg.rm(unpinned; mode=Pkg.PKGMODE_MANIFEST)'
```

**Review by omus — APPROVED** [review]
Looks good. You're changes build upon the changes from #7 so feel free to take your changes when resolving the conflict

**Pull request merged**